finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/ExistentialStability.scala

```diff
@@ -20,33 +20,46 @@ object ExistentialStability {
      * will update.
      */
     def stabilizeExistence: Var[Option[Var[T]]] = {
-      val init = unstable.sample().map(Var(_))
-      Var.async[Option[VarUp[T]]](init) { update =>
-        // the current inner Var, null if the outer Var is None
-        @volatile var current: VarUp[T] = null
-        @volatile var exists = false
-        val mu = new {}
+
+      // The inner Var is null until the first time unstable becomes a Some.
+      @volatile var inner: Var[T] = null
+
+      def makeInner(init: T) = Var.async(init) { update =>
         unstable.changes.respond {
-          case Some(t) if current == null => mu.synchronized {
-            // T created
-            exists = true
-            current = Var(t)
-            update() = Some(current)
-          }
-          case Some(t) if !exists => mu.synchronized {
-            // T re-created
-            exists = true
-            current() = t
-            update() = Some(current)
-          }
-          case Some(t) => mu.synchronized {
-            // T modified
-            current() = t
-          }
-          case None => mu.synchronized {
-            // T deleted
-            exists = false
-            update() = None
+          case Some(t) => update() = t
+          case _ =>
+        }
+      }
+
+      // Initial value for outer Var.
+      val init = unstable.sample().map { t =>
+        if (inner == null) {
+          inner = makeInner(t)
+        }
+        inner
+      }
+
+      @volatile var exists = init.isDefined
+      val mu = new {}
+
+      Var.async[Option[Var[T]]](init) { update =>
+        unstable.changes.respond { state =>
+          mu.synchronized {
```

Comment:
I notice there's a change in this branch from synchronizing each match arm separately to synchronizing the whole match block...is that also a necessary part of this change? If so, it would be great to get a mention of that in the commit message for future reference...

Reply:
Good question.  With the synchronized blocks just within the arms there is a potential race condition where two threads could pick the same arm because the case matching happens outside of the synchronize block.  I never observed this race condition happening in practice but to be safe I expanded the synchronized block to make sure this can never happen.